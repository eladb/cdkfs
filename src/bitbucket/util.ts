

export function reduceRenderObject( obj:Record<string, any> ) {

  const result = Object.keys(obj).reduce( ( inp:Record<string, any>, key:string) => {

    if (obj[ key ] !== undefined) inp[ key ] = obj[ key ];

    return inp;
  }, {} );

  if ( Object.keys( result).length === 0 ) return undefined;

  return result;
}


export function reduceRenderArray( arr:any[]|undefined ) {

  if ( arr === undefined ) return arr;
  else if (arr.length === 0) return undefined;
  return arr;
}