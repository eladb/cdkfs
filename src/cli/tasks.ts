import * as yargs from 'yargs';
import * as logging from '../logging';
import { TaskRuntime } from '../tasks';

/**
 * Reads .projen/tasks.json and adds CLI commands for all tasks.
 * @param ya yargs
 */
export function discoverTaskCommands(runtime: TaskRuntime, ya: yargs.Argv) {
  const tasks = runtime.manifest.tasks ?? {};
  for (const task of Object.values(tasks)) {
    ya.command(task.name, task.description ?? '', taskCommandHandler(task.name));
  }

  function taskCommandHandler(taskName: string) {
    return (args: yargs.Argv) => {
      args.option('inspect', { alias: 'i', desc: 'show all steps in this task' });

      const argv = args.argv;

      if (argv.inspect) {
        return inspectTask(taskName);
      } else {
        try {
          runtime.runTask(taskName);
        } catch (e) {
          logging.error(e.message);
          process.exit(1);
        }
      }
    };
  }

  function inspectTask(name: string, indent = 0) {
    const writeln = (s: string) => console.log(' '.repeat(indent) + s);

    const task = runtime.tryFindTask(name);
    if (!task) {
      throw new Error(`${name}: unable to resolve subtask with name "${name}"`);
    }

    for (const step of task.steps ?? []) {
      if (step.spawn) {
        writeln(`${step.spawn}:`);
        inspectTask(step.spawn, indent + 3);
      } else if (step.exec) {
        writeln(step.exec);
      }
    }
  };
}