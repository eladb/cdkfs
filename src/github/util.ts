export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

/**
 * Bash if equivalent execution of scripts that is compatible with shx.
 * Execution will fail, if the when or then command fails.
 *
 * @param when The condition to evaluate. Usually a `test` command. This command may be executed twice. Make sure it is non mutating.
 * @param then The script to run if the condition is true.
 * @param otherwise The script to run if the condition is false.
 */
export function shxIf(when: string, then: string, otherwise?: string): string {
  // Connect the condition and the 'then' command with a logical AND.
  // If the condition is true, this will execute the 'then' command.
  // This case is equivalent to TRUE && then()
  // If the condition is false, this will exit with a non-zero status code
  // (indicating an error) without executing the 'then' command.
  // This case is equivalent to FALSE && then()
  const if_clause = `${when} && (${then})`;

  // For the else clause, we need to check the 'when' condition again,
  // because the if clause can also return FALSE if the 'then' command fails.
  // However in this situation we don't want to execute the else clause but report an error.
  const else_clause = `${when} || (${otherwise})`;

  // Try the if_clause first. If it fails try the else clause.
  if (otherwise) {
    return `(${if_clause}) || (${else_clause})`;
  }

  // Just run the if clause
  return if_clause;
}
