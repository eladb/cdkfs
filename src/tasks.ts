import * as path from 'path';
import { PROJEN_DIR } from './common';
import { Component } from './component';
import { JsonFile } from './json';
import { Project } from './project';
import { Task, TaskOptions } from './task';
import { TasksManifest, TaskSpec } from './task-model';

/**
 * Defines project tasks.
 *
 * Tasks extend the projen CLI by adding subcommands to it. Task definitions are
 * synthesized into `.projen/tasks.json`.
 */
export class Tasks extends Component {
  /**
   * The project-relative path of the tasks manifest file.
   */
  public static readonly MANIFEST_FILE = path.posix.join(PROJEN_DIR, 'tasks.json');

  private readonly _tasks: { [name: string]: Task };
  private readonly _env: { [name: string]: string };

  constructor(project: Project) {
    super(project);

    const manifestFile = Tasks.MANIFEST_FILE;
    this._tasks = {};
    this._env = {};

    new JsonFile(project, manifestFile, {
      omitEmpty: true,
      obj: {
        tasks: (() => this.renderTasks()) as any,
        env: (() => this._env) as any,
      } as TasksManifest,
    });
  }

  /**
   * All tasks.
   */
  public get all() {
    return Object.values(this._tasks);
  }

  /**
   * Adds a task to a project.
   * @param name The name of the task
   * @param options Task options.
   */
  public addTask(name: string, options: TaskOptions = {}) {
    const task = new Task(name, options);
    if (this._tasks[name]) {
      throw new Error(`A task with the name ${name} already exists. To override it, call removeTask first and then create the new task.`);
    }
    this._tasks[name] = task;
    return task;
  }

  /**
   * Removes a task from a project.
   *
   * @param name The name of the task to remove.
   *
   * @returns The `Task` that was removed, otherwise `undefined`.
   */
  public removeTask(name: string): undefined | Task {
    const dependentTasks = this.all.filter(task => task.steps.find(step => step.spawn == name));
    if (dependentTasks.length > 0) {
      const errList = dependentTasks.map(depTask => depTask.name).join(', ');
      throw new Error(`Unable to remove task "${name}" because the following tasks depend on it: ${errList}`);
    }

    const task = this._tasks[name];
    if (task) {
      delete this._tasks[name];
      return task;
    } else {
      return undefined;
    }
  }

  /**
   * Adds global environment.
   * @param name Environment variable name
   * @param value Value
   */
  public addEnvironment(name: string, value: string) {
    this._env[name] = value;
  }

  /**
   * Returns a copy of the currently global environment for this project.
   */
  public get env(): { [key: string]: string } {
    return {
      ...this._env,
    };
  }

  /**
   * Finds a task by name. Returns `undefined` if the task cannot be found.
   * @param name The name of the task
   */
  public tryFind(name: string): undefined | Task {
    return this._tasks[name];
  }

  private renderTasks() {
    const tasks: { [name: string]: TaskSpec } = {};
    for (const task of Object.values(this._tasks).sort((x, y) => x.name.localeCompare(y.name))) {
      tasks[task.name] = task._renderSpec();
    }

    return tasks;
  }
}
