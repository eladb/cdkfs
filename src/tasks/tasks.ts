import * as path from 'path';
import { PROJEN_DIR } from '../common';
import { Component } from '../component';
import { JsonFile } from '../json';
import { Project } from '../project';
import { TasksManifest, TaskSpec } from './model';
import { Task, TaskOptions } from './task';

/**
 * Defines project tasks.
 *
 * Tasks extend the projen CLI by adding subcommands to it. Task definitions are
 * synthesized into `.projen/tasks.json`.
 */
export class Tasks extends Component {
  /**
   * The project-relative path of the tasks manifest file.
   */
  public static readonly MANIFEST_FILE = path.posix.join(PROJEN_DIR, 'tasks.json');

  private readonly _tasks: { [name: string]: Task };
  private readonly _env: { [name: string]: string };

  constructor(project: Project) {
    super(project);

    const manifestFile = Tasks.MANIFEST_FILE;
    this._tasks = {};
    this._env = {};

    new JsonFile(project, manifestFile, {
      marker: true,
      omitEmpty: true,
      obj: {
        tasks: (() => this.renderTasks()) as any,
        env: (() => this._env) as any,
      } as TasksManifest,
    });
  }

  /**
   * All tasks.
   */
  public get all() {
    return Object.values(this._tasks);
  }

  /**
   * Adds a task to a project.
   * @param name The name of the task
   * @param options Task options.
   */
  public addTask(name: string, options: TaskOptions = {}) {
    const task = new Task(this, name, options);
    this._tasks[name] = task;
    return task;
  }

  /**
   * Adds global environment.
   * @param name Environment variable name
   * @param value Value
   */
  public addEnvironment(name: string, value: string) {
    this._env[name] = value;
  }

  /**
   * Returns a copy of the currently global environment for this project.
   */
  public get env(): { [key: string]: string } {
    return {
      ...this._env,
    };
  }

  /**
   * Finds a task by name. Returns `undefined` if the task cannot be found.
   * @param name The name of the task
   */
  public tryFind(name: string): undefined | Task {
    return this._tasks[name];
  }

  private renderTasks() {
    const tasks: { [name: string]: TaskSpec } = {};
    for (const task of Object.values(this._tasks)) {
      tasks[task.name] = task._renderSpec();
    }

    return tasks;
  }
}
