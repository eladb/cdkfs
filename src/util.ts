import * as child_process from 'child_process';
import * as path from 'path';
import * as fs from 'fs-extra';
import * as logging from './logging';
// eslint-disable-next-line @typescript-eslint/no-require-imports
const decamelize = require('decamelize');

export function exec(command: string, options?: child_process.ExecSyncOptions) {
  logging.verbose(command);
  return child_process.execSync(command, {
    stdio: ['inherit', process.stderr, 'pipe'],
    ...options,
  });
}

/**
 * Executes `command` and returns its value or undefined if the command failed.
 */
export function execOrUndefined(command: string, options?: child_process.ExecSyncOptions): string | undefined {
  try {
    const value = child_process.execSync(command, { stdio: ['inherit', 'pipe', 'ignore'], ...options }).toString('utf-8').trim();
    if (!value) { return undefined; } // an empty string is the same as undefined
    return value;
  } catch {
    return undefined;
  }
}

export interface WriteFileOptions {
  /**
   * Whether the generated file should be marked as executable.
   *
   * @default false
   */
  executable?: boolean;

  /**
   * Whether the generated file should be readonly.
   *
   * @default false
   */
  readonly?: boolean;
}

export function getFilePermissions(options: WriteFileOptions): string {
  const readonly = options.readonly ?? false;
  const executable = options.executable ?? false;
  if (readonly && executable) {
    return '500';
  } else if (readonly) {
    return '400';
  } else if (executable) {
    return '755';
  } else {
    return '644';
  }
}

export function writeFile(filePath: string, data: any, options: WriteFileOptions = {}) {
  if (fs.existsSync(filePath)) {
    fs.chmodSync(filePath, '600');
  }

  fs.mkdirpSync(path.dirname(filePath));
  fs.writeFileSync(filePath, data);

  fs.chmodSync(filePath, getFilePermissions(options));
}

/**
 * Decamelizes the keys of an object structure, recursing through child objects and arrays.
 * @experimental
 */
export interface DecamelizeRecursivelyOptions {
  /**
   * Max depth to recurse before erroring.
   * @default 10
   */
  maxDepth?: number;

  /**
   * Returns true when a key should be decamelized
   * @default - all keys are decamelized
   */
  shouldDecamelize?: (path: string[], value: any) => boolean;

  /**
   * Separator for decamelizing.
   * @default "_"
   */
  separator?: string;

  /**
   * Current path.
   * @internal
   */
  path?: string[];
}

export function decamelizeKeysRecursively(input: any, opt?: DecamelizeRecursivelyOptions): any {
  const shouldAlwaysDecamelize = () => true;
  const shouldDecamelize = opt?.shouldDecamelize ?? shouldAlwaysDecamelize;
  const separator = opt?.separator ?? '_';
  const path_ = opt?.path ?? [];
  const maxDepth = opt?.maxDepth ?? 10;

  if (path_.length > maxDepth) {
    throw new Error('Decamelled too deeply - check that the input has no circular references');
  }

  if (Array.isArray(input)) {
    return input.map((k, i) => decamelizeKeysRecursively(k, {
      ...opt,
      path: [...path_, i.toString()],
    }));
  }

  if (typeof input === 'object' && input !== null) {
    const mappedObject: Record<string, any> = {};
    for (const [key, value] of Object.entries(input)) {
      const transformedKey = shouldDecamelize([...path_, key], value)
        ? decamelize(key, separator)
        : key;

      mappedObject[transformedKey] = decamelizeKeysRecursively(value, {
        ...opt,
        path: [...path_, key],
      });
    }

    return mappedObject;
  }

  return input;
}

/**
 * Returns false if value is unset or a falsey value, and true otherwise.
 * @param value an environment variable
 */
export function isTruthy(value: string | undefined): boolean {
  return !(value === undefined || ['null', 'undefined', '0', 'false', ''].includes(value.toLocaleLowerCase()));
}

/**
 * Type of a map mapping strings to some arbitrary type
 */
export type Obj<T> = { [key: string]: T };

/**
 * Return whether the given value is an object
 *
 * Even though arrays and instances of classes technically are objects, we
 * usually want to treat them differently, so we return false in those cases.
 */
export function isObject(x: any): x is Obj<any> {
  return x !== null && typeof x === 'object' && !Array.isArray(x)
    && x.constructor.name === 'Object';
}

/**
 * Recursively merge objects together
 *
 * The leftmost object is mutated and returned. Arrays are not merged
 * but overwritten just like scalars.
 *
 * If an object is merged into a non-object, the non-object is lost.
 *
 * `undefined`s will cause a value to be deleted if destructive is enabled.
 */
export function deepMerge(objects: Array<Obj<any> | undefined>, destructive: boolean = false) {
  function mergeOne(target: Obj<any>, source: Obj<any>) {
    for (const key of Object.keys(source)) {
      const value = source[key];

      if (isObject(value)) {
        // if the value at the target is not an object, override it with an
        // object so we can continue the recursion
        if (!isObject(target[key])) {
          target[key] = value;
        }
        mergeOne(target[key], value);

        // if the result of the merge is an empty object, it's because the
        // eventual value we assigned is `undefined`, and there are no
        // sibling concrete values alongside, so we can delete this tree.
        const output = target[key];
        if (isObject(output) && Object.keys(output).length === 0 && destructive) {
          delete target[key];
        }
      } else if (value === undefined && destructive) {
        delete target[key];
      } else if (typeof value !== 'undefined') {
        target[key] = value;
      }
    }
  }

  const others = objects.filter(x => x != null) as Array<Obj<any>>;

  if (others.length === 0) { return {}; }
  const into = others.splice(0, 1)[0];

  others.forEach(other => mergeOne(into, other));
  return into;
}

/*
 * Deduplicate values in a list, returning a new array.
 * @param array list of values
 */
export function dedupArray<T>(array: T[]): T[] {
  return array.filter((val, idx) => array.indexOf(val) === idx);
}

/**
 * Returns a sorted version of `x` or `undefined` if it is an empty array or object.
 */
export function sorted<T>(x: T) {
  if (x == null) { return undefined; }
  if (Array.isArray(x)) {
    if (x.length === 0) { return undefined; }
    return (x as unknown[]).sort();
  } else if (typeof x === 'object') {
    if (Object.keys(x).length === 0) { return undefined; }
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(x).sort(([l], [r]) => l.localeCompare(r))) {
      result[key] = value;
    }
    return result as T;
  } else {
    return x;
  }
}

export function formatAsPythonModule(name: string) {
  return name.replace(/-/g, '_').replace(/\./g, '_');
}
