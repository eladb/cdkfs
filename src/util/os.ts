/**
 * Provides information about the current operating system.
 */
export class OsInspector {
  /**
   * Whether the current OS is AIX.
   */
  public readonly isAix: boolean;

  /**
   * Whether the current OS is FreeBSD.
   */
  public readonly isFreebsd: boolean;

  /**
   * Whether the current OS is Linux.
   */
  public readonly isLinux: boolean;

  /**
   * Whether the current OS is MacOS.
   */
  public readonly isMacos: boolean;

  /**
   * Whether the current OS is OpenBSD.
   */
  public readonly isOpenbsd: boolean;

  /**
   * Whether the current OS is SunOS.
   */
  public readonly isSunos: boolean;

  /**
   * Whether the current OS is Windows.
   */
  public readonly isWindows: boolean;

  /**
   * Contains a list of file extensions that the current Windows installation considers to be executable files
   */
  private _pathExt?: Set<string>;

  constructor() {
    this.isAix = process.platform === "aix";
    this.isWindows = process.platform === "win32";
    this.isMacos = process.platform === "darwin";
    this.isFreebsd = process.platform === "freebsd";
    this.isLinux = process.platform === "linux";
    this.isOpenbsd = process.platform === "openbsd";
    this.isSunos = process.platform === "sunos";
  }

  public get pathExt(): Set<string> {
    if (!this.isWindows) {
      throw new Error("pathExt is only available on Windows");
    }

    if (!this._pathExt) {
      this._pathExt = new Set(
        process.env.PATHEXT?.split(";").map((extension) =>
          extension.toLowerCase()
        )
      );
    }

    return this._pathExt;
  }
}
