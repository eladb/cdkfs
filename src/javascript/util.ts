import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve } from "path";
import * as semver from "semver";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/.]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
