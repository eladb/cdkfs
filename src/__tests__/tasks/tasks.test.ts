import { spawnSync } from 'child_process';
import { EOL } from 'os';
import { Project } from '../..';
import { Task, Tasks, TasksManifest, TaskStep } from '../../tasks';
import { TestProject, synthSnapshot } from '../util';

test('no tasks, no tasks file', () => {
  const p = new TestProject();
  expect(synthTasksManifest(p)).toBeUndefined();
});

test('empty task', () => {
  const p = new TestProject();

  // WHEN
  p.addTask('empty');

  // THEN
  expectManifest(p, {
    tasks: {
      empty: {
        name: 'empty',
      },
    },
  });
});

test('multiple "exec" commands', () => {
  const p = new TestProject();

  // WHEN
  const task = p.addTask('hello', {
    description: 'hello, world',
    exec: 'echo hello', // initial command
    env: {
      FOO: 'bar',
    },
  });

  task.exec('echo world');
  task.exec('echo "with quotes"');
  task.env('BAR', 'baz');

  // THEN
  expectManifest(p, {
    tasks: {
      hello: {
        name: 'hello',
        description: 'hello, world',
        env: {
          FOO: 'bar',
          BAR: 'baz',
        },
        steps: [
          { exec: 'echo hello' },
          { exec: 'echo world' },
          { exec: 'echo "with quotes"' },
        ],
      },
    },
  });
});

test('subtasks', () => {
  // GIVEN
  const p = new TestProject();
  const hello = p.addTask('hello', { exec: 'echo hello' });
  const world = p.addTask('world');

  // WHEN
  world.exec('echo "running hello"');
  world.spawn(hello);

  // THEN
  expectManifest(p, {
    tasks: {
      hello: {
        name: 'hello',
        steps: [{ exec: 'echo hello' }],
      },
      world: {
        name: 'world',
        steps: [
          { exec: 'echo "running hello"' },
          { spawn: 'hello' },
        ],
      },
    },
  });
});

test('reset() can be used to reset task steps', () => {
  // GIVEN
  const p = new TestProject();
  const t0 = p.addTask('your-task');
  const t = p.addTask('my-task');
  t.exec('line1');
  t.spawn(t0);
  t.exec('line2');

  // WHEN
  t.reset('line3');
  t.exec('line4');

  // THEN
  expectManifest(p, {
    tasks: {
      'your-task': {
        name: 'your-task',
      },
      'my-task': {
        name: 'my-task',
        steps: [
          { exec: 'line3' },
          { exec: 'line4' },
        ],
      },
    },
  });
});

test('prependXXX() can be used to add steps from the top', () => {
  // GIVEN
  const p = new TestProject();
  const sub = p.addTask('my-sub-task', { exec: 'subexec' });

  const t = p.addTask('my-task');
  t.exec('line1');

  // WHEN
  t.prependExec('line2');
  t.prependSpawn(sub);
  t.prependSay('message');

  // THEN
  expectManifest(p, {
    tasks: {
      'my-sub-task': {
        name: 'my-sub-task',
        steps: [
          { exec: 'subexec' },
        ],
      },
      'my-task': {
        name: 'my-task',
        steps: [
          { say: 'message' },
          { spawn: 'my-sub-task' },
          { exec: 'line2' },
          { exec: 'line1' },
        ],
      },
    },
  });
});

test('env() can be used to add environment variables', () => {
  // GIVEN
  const p = new TestProject();
  const t = p.addTask('my-task', {
    env: {
      INITIAL: '123',
      ENV: '456',
    },
  });

  // WHEN
  t.env('FOO', 'BAR');
  t.env('HELLO', 'world');

  // THEN
  expectManifest(p, {
    tasks: {
      'my-task': {
        name: 'my-task',
        env: {
          INITIAL: '123',
          ENV: '456',
          FOO: 'BAR',
          HELLO: 'world',
        },
      },
    },
  });
});

test('.steps can be used to list all steps in the current task', () => {
  // GIVEN
  const p = new TestProject();
  const t0 = p.addTask('your');
  const t = p.addTask('my');
  t.exec('step1');
  t.exec('step2');
  t.exec('step3');
  t.spawn(t0);
  t.exec('step4');

  // WHEN
  const steps = t.steps;

  // THEN
  expect(steps).toStrictEqual([
    { exec: 'step1' },
    { exec: 'step2' },
    { exec: 'step3' },
    { spawn: 'your' },
    { exec: 'step4' },
  ] as TaskStep[]);
});

test('"condition" can be used to define a command that will determine if a task should be skipped', () => {
  // GIVEN
  const p = new TestProject();
  p.addTask('foo', {
    condition: 'false',
    exec: 'foo bar',
  });

  // THEN
  expectManifest(p, {
    tasks: {
      foo: {
        name: 'foo',
        condition: 'false',
        steps: [
          { exec: 'foo bar' },
        ],
      },
    },
  });
});

describe('toShellCommand()', () => {

  test('single step', () => {
    // GIVEN
    const p = new TestProject();

    // WHEN
    const t = p.addTask('foo', { exec: 'echo hi there' });

    // THEN
    expect(t.toShellCommand()).toMatchSnapshot();
    expect(shell(t)).toStrictEqual(['hi there']);
  });

  test('with a name', () => {
    // GIVEN
    const p = new TestProject();
    const t = p.addTask('foo');

    // WHEN
    t.exec('echo step 1', { name: 'STEP I' });

    // THEN
    expect(t.toShellCommand()).toMatchSnapshot();
    expect(shell(t)).toStrictEqual([
      'STEP I',
      'step 1',
    ]);
  });

  test('with a condition', () => {
    // GIVEN
    const p = new TestProject();
    const t = p.addTask('foo', {
      condition: '[ "${RUNME}" = "1" ]',
    });

    // WHEN
    t.exec('echo hello, world', { name: 'STEP I' });

    // THEN
    expect(t.toShellCommand()).toMatchSnapshot();
    expect(shell(t)).toStrictEqual([]);
    expect(shell(t, { RUNME: '1' })).toStrictEqual([
      'STEP I',
      'hello, world',
    ]);
  });

  test('multiple steps', () => {
    // GIVEN
    const p = new TestProject();
    const t = p.addTask('foo');

    // WHEN
    t.exec('echo step 1', { name: 'STEP I' });
    t.exec('echo step 2');
    t.exec('echo step 3', { name: 'STEP III' });

    // THEN
    expect(t.toShellCommand()).toMatchSnapshot();
    expect(shell(t)).toStrictEqual([
      'STEP I',
      'step 1',
      'step 2',
      'STEP III',
      'step 3',
    ]);
  });

  test('subtasks', () => {
    // GIVEN
    const p = new TestProject();
    const t1 = p.addTask('t1');
    const t2 = p.addTask('t2');

    // WHEN
    t1.exec('echo task1-step1');
    t1.exec('echo task1-step2');
    t2.exec('echo task2-step1');
    t2.spawn(t1, { name: 'spawning t1' });
    t2.exec('echo task2-step3');

    // THEN
    expect(t2.toShellCommand()).toMatchSnapshot();
    expect(shell(t2)).toStrictEqual([
      'task2-step1',
      'spawning t1',
      'task1-step1',
      'task1-step2',
      'task2-step3',
    ]);
  });

  test('with environment', () => {
    // GIVEN
    const p = new TestProject();
    p.tasks.addEnvironment('FOO', 'hello'); // global environment

    // WHEN
    const t1 = p.addTask('t1', {
      env: {
        BAR: '$(echo world)', // local environment
      },
      exec: 'echo $FOO, $BAR !',
    });

    // THEN
    expect(t1.toShellCommand()).toMatchSnapshot();
    expect(shell(t1)).toStrictEqual([
      'hello, world !',
    ]);
  });
});

function shell(t: Task, env: { [k: string]: string } = {}) {
  const result = spawnSync(t.toShellCommand(), { shell: true, env: { ...process.env, ...env } });
  if (result.status !== 0) {
    throw new Error(`non-zero exit code ${result.status}: ${result.stderr.toString('utf-8')}`);
  }

  return result.stdout.toString('utf-8').trim().split(EOL).filter(x => x).map(x => x.trim());
}

function expectManifest(p: Project, toStrictEqual: TasksManifest) {
  const manifest = synthTasksManifest(p);
  delete manifest['//'];
  expect(manifest).toStrictEqual(toStrictEqual);
}

function synthTasksManifest(p: Project) {
  return synthSnapshot(p)[Tasks.MANIFEST_FILE];;
}