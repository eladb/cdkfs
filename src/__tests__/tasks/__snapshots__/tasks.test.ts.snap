// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toShellCommand() multiple steps 1`] = `"(  ( echo STEP I ) && ( echo step 1 ) && ( echo step 2 ) && ( echo STEP III ) && ( echo step 3 ) )"`;

exports[`toShellCommand() single step 1`] = `"(  ( echo hi there ) )"`;

exports[`toShellCommand() subtasks 1`] = `"(  ( echo task2-step1 ) && ( echo spawning t1 ) && ( ( (  ( echo task1-step1 ) && ( echo task1-step2 ) ) ) ) && ( echo task2-step3 ) )"`;

exports[`toShellCommand() with a condition 1`] = `"(  ! ( [ \\"\${RUNME}\\" = \\"1\\" ] ) || ( ( echo STEP I ) && ( echo hello, world ) ) )"`;

exports[`toShellCommand() with a name 1`] = `"(  ( echo STEP I ) && ( echo step 1 ) )"`;

exports[`toShellCommand() with environment 1`] = `"( FOO=\\"hello\\"; BAR=\\"$(echo world)\\";  ( echo $FOO, $BAR ! ) )"`;
